// @workbench-stack/core/src/interfaces/view-NOTUSED.ts
import { Command } from '@vscode-alt/monaco-editor/esm/vs/editor/common/modes';
import { UriComponents } from '@vscode-alt/monaco-editor/esm/vs/base/common/uri';

// src/vs/workbench/common/views.ts


export enum TreeItemCollapsibleState {
    None = 0,
    Collapsed = 1,
    Expanded = 2,
}

export interface ITreeItemLabel {

    label: string;

    highlights?: [number, number][];

}

export interface IRevealOptions {

    select?: boolean;

    focus?: boolean;

    expand?: boolean | number;

}

export interface ThemeIcon {
    readonly id: string;
}

export interface ITreeItem {

    handle: string;

    parentHandle?: string;

    collapsibleState: TreeItemCollapsibleState;

    label?: ITreeItemLabel;

    description?: string | boolean;

    icon?: UriComponents;

    iconDark?: UriComponents;

    themeIcon?: ThemeIcon;

    resourceUri?: UriComponents;

    tooltip?: string;

    contextValue?: string;

    command?: Command;

    children?: ITreeItem[];
}

export interface ITreeViewDataProvider {

    getChildren(element?: ITreeItem): Promise<ITreeItem[]>;

}
