// @workbench-stack/core/src/interfaces/theme/color-registry.ts
import { Color } from '@vscode-alt/monaco-editor/esm/vs/base/common/color';


export interface ThemeColor {
    id: string;
}
export type ColorIdentifier = string;


export const DARK: ThemeType = 'dark';
export const LIGHT: ThemeType = 'light';
export const HIGH_CONTRAST: ThemeType = 'hc';
export type ThemeType = 'light' | 'dark' | 'hc';

export interface ITheme {
    readonly type: ThemeType;

    /**
     * Resolves the color of the given color identifier. If the theme does not
     * specify the color, the default color is returned unless <code>useDefault</code> is set to false.
     * @param color the id of the color
     * @param useDefault specifies if the default color should be used. If not set, the default is used.
     */
    getColor(color: ColorIdentifier, useDefault?: boolean): Color | null;

    /**
     * Returns whether the theme defines a value for the color. If not, that means the
     * default color will be used.
     */
    defines(color: ColorIdentifier): boolean;

    // @sri custom addition
    name: string;
    className?: string;
    text?: string;
    defaultSettings?: any;
}
