// @files-stack/core/lib/constants/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHANGES_LENGTH_LIMIT = exports.CONTENT_SIZE_LIMIT = exports.MAX_HEAP_SIZE = exports.MAX_FILE_SIZE = exports.ETAG_DISABLED = exports.FALLBACK_MAX_MEMORY_SIZE_MB = exports.MIN_MAX_MEMORY_SIZE_MB = exports.FILES_EXCLUDE_CONFIG = exports.FILES_ASSOCIATIONS_CONFIG = exports.CONTENT_CHANGE_EVENT_BUFFER_DELAY = exports.PubsubIdentifier = exports.MOLECULER_KUBE_TEXT_FILESERVICE = exports.MOLECULER_KUBE_FILESERVICE = exports.NATS_HEMERA_KUBE_TEXT_FILESERVICE = exports.NATS_HEMERA_KUBE_FILESERVICE = void 0;
/* tslint:disable */
exports.NATS_HEMERA_KUBE_FILESERVICE = 'NATS_HEMERA_KUBE_FILESERVICE';
exports.NATS_HEMERA_KUBE_TEXT_FILESERVICE = 'NATS_HEMERA_KUBE_TEXT_FILESERVICE';
exports.MOLECULER_KUBE_FILESERVICE = 'MOLECULER_KUBE_FILESERVICE';
exports.MOLECULER_KUBE_TEXT_FILESERVICE = 'MOLECULER_KUBE_TEXT_FILESERVICE';
var PubsubIdentifier;
(function (PubsubIdentifier) {
    PubsubIdentifier["FILE_UPDATED"] = "FILE_UPDATED";
    PubsubIdentifier["FILE_OPERATION"] = "FILE_OPERATION";
    PubsubIdentifier["FILE_CONTENT_STREAM"] = "FILE_CONTENT_STREAM";
    PubsubIdentifier["onWillActivateFileSystemProvider"] = "onWillActivateFileSystemProvider";
    PubsubIdentifier["onDidChangeFileSystemProviderRegistrations"] = "onDidChangeFileSystemProviderRegistrations,";
})(PubsubIdentifier = exports.PubsubIdentifier || (exports.PubsubIdentifier = {}));
// src/vs/platform/files/common/files.ts
exports.CONTENT_CHANGE_EVENT_BUFFER_DELAY = 1000;
exports.FILES_ASSOCIATIONS_CONFIG = 'files.associations';
exports.FILES_EXCLUDE_CONFIG = 'files.exclude';
exports.MIN_MAX_MEMORY_SIZE_MB = 2048;
exports.FALLBACK_MAX_MEMORY_SIZE_MB = 4096;
/**
 * A hint to disable etag checking for reading/writing.
 */
exports.ETAG_DISABLED = '';
// See https://github.com/Microsoft/vscode/issues/30180
const WIN32_MAX_FILE_SIZE = 300 * 1024 * 1024; // 300 MB
const GENERAL_MAX_FILE_SIZE = 16 * 1024 * 1024 * 1024; // 16 GB
// See https://github.com/v8/v8/blob/5918a23a3d571b9625e5cce246bdd5b46ff7cd8b/src/heap/heap.cc#L149
const WIN32_MAX_HEAP_SIZE = 700 * 1024 * 1024; // 700 MB
const GENERAL_MAX_HEAP_SIZE = 700 * 2 * 1024 * 1024; // 1400 MB
exports.MAX_FILE_SIZE = process.arch === 'ia32' ? WIN32_MAX_FILE_SIZE : GENERAL_MAX_FILE_SIZE;
exports.MAX_HEAP_SIZE = process.arch === 'ia32' ? WIN32_MAX_HEAP_SIZE : GENERAL_MAX_HEAP_SIZE;
// @sri our limits
exports.CONTENT_SIZE_LIMIT = 819 * 1024; // 80% of 1MB in byte
exports.CHANGES_LENGTH_LIMIT = 100 * 1024; // 100KB
