// @workbench-stack/core/src/core/editor/editor-input.ts

import { 
    BINARY_DIFF_EDITOR_ID,
    TEXT_DIFF_EDITOR_ID,
    EDITOR_INPUT_ID,
    IEditorInput,
    ISideBySideEditorInput,
    IDiffEditorInput,
    IEDITOR_INPUT_ID,
} from '../../interfaces'
import { CacheTypenames } from '../../constants';
import { URI } from 'cdecode/client';

export function generateSideBySideEditorInput(name, description, _details: IEditorInput, _master: IEditorInput): ISideBySideEditorInput {

    return {
        label: name,
        master: _master,
        details: _details,
        typeId: EDITOR_INPUT_ID.SIDE_BY_SIDE_EDITOR_INPUT,
        description: description,
        __typename: CacheTypenames.SideBySideEditorInput,
    }
}

export function generateDiffEditorInput(name: string, description: string | null, original: IEditorInput, modified: IEditorInput, forceOpenAsBinary?: boolean): IDiffEditorInput {

    return {
        label: name,
        typeId: EDITOR_INPUT_ID.DIFF_EDITOR_INPUT,
        originalInput: original,
        modifiedInput: modified,
        preferredEditorId: forceOpenAsBinary ? BINARY_DIFF_EDITOR_ID : TEXT_DIFF_EDITOR_ID,
        __typename: CacheTypenames.DiffEditorInput,
    }
}


export function generateResourceEditorInput(name: string, description: string | null, resource: URI, preferredMode: string | undefined ) {

    return {
        name,
        description,
        resource,
        __typename: CacheTypenames.ResourceEditorInput,
    }
}
