// @workbench-stack/core/lib/interfaces/editor/editor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EncodingMode = exports.ConfirmResult = exports.Verbosity = exports.BINARY_DIFF_EDITOR_ID = exports.TEXT_DIFF_EDITOR_ID = void 0;
// copied from vs/workbench/common/editor.ts
/**
 * Text diff editor id.
 */
exports.TEXT_DIFF_EDITOR_ID = 'workbench.editors.textDiffEditor';
/**
 * Binary diff editor id.
 */
exports.BINARY_DIFF_EDITOR_ID = 'workbench.editors.binaryResourceDiffEditor';
var Verbosity;
(function (Verbosity) {
    Verbosity[Verbosity["SHORT"] = 0] = "SHORT";
    Verbosity[Verbosity["MEDIUM"] = 1] = "MEDIUM";
    Verbosity[Verbosity["LONG"] = 2] = "LONG";
})(Verbosity = exports.Verbosity || (exports.Verbosity = {}));
var ConfirmResult;
(function (ConfirmResult) {
    ConfirmResult[ConfirmResult["SAVE"] = 0] = "SAVE";
    ConfirmResult[ConfirmResult["DONT_SAVE"] = 1] = "DONT_SAVE";
    ConfirmResult[ConfirmResult["CANCEL"] = 2] = "CANCEL";
})(ConfirmResult = exports.ConfirmResult || (exports.ConfirmResult = {}));
var EncodingMode;
(function (EncodingMode) {
    /**
     * Instructs the encoding support to encode the current input with the provided encoding
     */
    EncodingMode[EncodingMode["Encode"] = 0] = "Encode";
    /**
     * Instructs the encoding support to decode the current input with the provided encoding
     */
    EncodingMode[EncodingMode["Decode"] = 1] = "Decode";
})(EncodingMode = exports.EncodingMode || (exports.EncodingMode = {}));
// remove in future
// export interface ServicesAccessor {
// 	get<T>(id: ServiceIdentifier<T>, isOptional?: any): T;
// }
