// @workbench-stack/core/lib/constants/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SETTINGS_EDITOR_2_INPUT = exports.StatusbarExtensions = exports.ClientExtensions = exports.KEYBINDING_EDITOR_ID = exports.BINARY_FILE_EDITOR_ID = exports.FILE_EDITOR_INPUT_ID = exports.TEXT_FILE_EDITOR_ID = exports.SETTINGS_EDITOR_ID = exports.PREFERENCE_EDITOR_ID = void 0;
exports.PREFERENCE_EDITOR_ID = 'workbench.editor.preferenceeditor';
exports.SETTINGS_EDITOR_ID = 'workbench.editor.settingseditor';
/**
 * Text file editor id.
 */
exports.TEXT_FILE_EDITOR_ID = 'workbench.editors.files.textFileEditor';
/**
 * File editor input id.
 */
exports.FILE_EDITOR_INPUT_ID = 'workbench.editors.files.fileEditorInput';
/**
 * Binary file editor id.
 */
exports.BINARY_FILE_EDITOR_ID = 'workbench.editors.files.binaryFileEditor';
exports.KEYBINDING_EDITOR_ID = 'workbench.editor.keybindingeditor';
exports.ClientExtensions = {
    Editors: 'workbench.contributions.editors',
    EditorInputFactories: 'workbench.contributions.editor.inputFactories',
};
exports.StatusbarExtensions = {
    Statusbar: 'workbench.contributions.statusbar',
};
exports.SETTINGS_EDITOR_2_INPUT = 'workbench.input.settings2';
