// @workbench-stack/core/lib/constants/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WORKSPACE_STANDALONE_CONFIGURATIONS = exports.LAUNCH_CONFIGURATION_KEY = exports.TASKS_CONFIGURATION_KEY = exports.launchSchemaId = exports.folderSettingsSchemaId = exports.workspaceSettingsSchemaId = exports.machineSettingsSchemaId = exports.userSettingsSchemaId = exports.defaultSettingsSchemaId = exports.FOLDER_SETTINGS_PATH = exports.FOLDER_SETTINGS_NAME = exports.FOLDER_CONFIG_FOLDER_NAME = void 0;
// export const CONFIG_DEFAULT_NAME = 'settings';
// export const WORKSPACE_CONFIG_FOLDER_DEFAULT_NAME = '.cdecode';
// export const WORKSPACE_CONFIG_DEFAULT_PATH = `${WORKSPACE_CONFIG_FOLDER_DEFAULT_NAME}/${CONFIG_DEFAULT_NAME}.json`;
const config_1 = require("../config");
exports.FOLDER_CONFIG_FOLDER_NAME = config_1.config.FOLDER_CONFIG_FOLDER_NAME;
exports.FOLDER_SETTINGS_NAME = config_1.config.FOLDER_SETTINGS_NAME;
exports.FOLDER_SETTINGS_PATH = `${exports.FOLDER_CONFIG_FOLDER_NAME}/${exports.FOLDER_SETTINGS_NAME}.json`;
exports.defaultSettingsSchemaId = 'vscode://schemas/settings/default';
exports.userSettingsSchemaId = 'vscode://schemas/settings/user';
exports.machineSettingsSchemaId = 'vscode://schemas/settings/machine';
exports.workspaceSettingsSchemaId = 'vscode://schemas/settings/workspace';
exports.folderSettingsSchemaId = 'vscode://schemas/settings/folder';
exports.launchSchemaId = 'vscode://schemas/launch';
exports.TASKS_CONFIGURATION_KEY = 'tasks';
exports.LAUNCH_CONFIGURATION_KEY = 'launch';
exports.WORKSPACE_STANDALONE_CONFIGURATIONS = Object.create(null);
exports.WORKSPACE_STANDALONE_CONFIGURATIONS[exports.TASKS_CONFIGURATION_KEY] = `${exports.FOLDER_CONFIG_FOLDER_NAME}/${exports.TASKS_CONFIGURATION_KEY}.json`;
exports.WORKSPACE_STANDALONE_CONFIGURATIONS[exports.LAUNCH_CONFIGURATION_KEY] = `${exports.FOLDER_CONFIG_FOLDER_NAME}/${exports.LAUNCH_CONFIGURATION_KEY}.json`;
