// @files-stack/core/lib/constants/graphql-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILE_STAT = exports.FileClientTypenames = exports.EDITOR_DEFAULT_HEIGHT = void 0;
exports.EDITOR_DEFAULT_HEIGHT = 80;
var FileClientTypenames;
(function (FileClientTypenames) {
    FileClientTypenames["FileStatWithMetadata"] = "FileStatWithMetadata";
    FileClientTypenames["FileStat"] = "FileStat";
    FileClientTypenames["TextFileContent"] = "TextFileContent";
    FileClientTypenames["FileContent"] = "FileContent";
    FileClientTypenames["StreamContent"] = "StreamContent";
    FileClientTypenames["ChangedContent"] = "ChangedContent";
    FileClientTypenames["ContentChangedEvent"] = "ContentChangedEvent";
    FileClientTypenames["ContentChangedEventJson"] = "ContentChangedEventJson";
    FileClientTypenames["CodeEditorViewState"] = "CodeEditorViewState";
    FileClientTypenames["FileEditorInput"] = "FileEditorInput";
})(FileClientTypenames = exports.FileClientTypenames || (exports.FileClientTypenames = {}));
// File Stat 
exports.FILE_STAT = 'FileStat';
