// @vscode-alt/monaco-editor/esm/vs/platform/contextkey/common/contextkeys.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
import { RawContextKey } from './contextkey.js';
import { isMacintosh, isLinux, isWindows, isWeb } from '../../../base/common/platform.js';
export const IsMacContext = new RawContextKey('isMac', isMacintosh);
export const IsLinuxContext = new RawContextKey('isLinux', isLinux);
export const IsWindowsContext = new RawContextKey('isWindows', isWindows);
export const IsWebContext = new RawContextKey('isWeb', isWeb);
export const IsMacNativeContext = new RawContextKey('isMacNative', isMacintosh && !isWeb);
export const IsDevelopmentContext = new RawContextKey('isDevelopment', false);
export const InputFocusedContextKey = 'inputFocus';
export const InputFocusedContext = new RawContextKey(InputFocusedContextKey, false);
