// @workbench-stack/core/lib/services/workspaces/abstract-workspace-context-service.js
"use strict";
/* tslint:disable */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractWorkspaceContextService = void 0;
const inversify_1 = require("inversify");
const resources_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/resources");
const core_1 = require("../../core");
const uri_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/uri");
const event_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/event");
const lifecycle_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/lifecycle");
const async_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/async");
const core_2 = require("@common-stack/core");
let AbstractWorkspaceContextService = class AbstractWorkspaceContextService extends lifecycle_1.Disposable {
    constructor(logger) {
        super();
        this._onDidChangeConfiguration = this._register(new event_1.Emitter());
        this.onDidChangeConfiguration = this._onDidChangeConfiguration.event;
        this._onDidChangeWorkspaceFolders = this._register(new event_1.Emitter());
        this.onDidChangeWorkspaceFolders = this._onDidChangeWorkspaceFolders.event;
        this._onDidChangeWorkspaceName = this._register(new event_1.Emitter());
        this.onDidChangeWorkspaceName = this._onDidChangeWorkspaceName.event;
        this._onDidChangeWorkbenchState = this._register(new event_1.Emitter());
        this.onDidChangeWorkbenchState = this._onDidChangeWorkbenchState.event;
        this.logger = logger.child({ className: 'AbstractWorkspaceService' });
        this.completeWorkspaceBarrier = new async_1.Barrier();
    }
    getCompleteWorkspace() {
        return this.completeWorkspaceBarrier.wait().then(() => this.getWorkspace());
    }
    getWorkspace() {
        return this.workspace;
    }
    getWorkbenchState() {
        // Workspace has configuration file
        if (this.workspace.configuration) {
            return 3 /* WORKSPACE */;
        }
        // Folder has single root
        if (this.workspace.folders.length === 1) {
            return 2 /* FOLDER */;
        }
        // Empty
        return 1 /* EMPTY */;
    }
    getWorkspaceFolder(resource) {
        return this.workspace.getFolder(resource);
    }
    isInsideWorkspace(resource) {
        return !!this.getWorkspaceFolder(resource);
    }
    isCurrentWorkspace(workspaceIdentifier) {
        switch (this.getWorkbenchState()) {
            case 2 /* FOLDER */:
                return core_1.isSingleFolderWorkspaceIdentifier(workspaceIdentifier) && resources_1.isEqual(workspaceIdentifier, this.workspace.folders[0].uri);
            case 3 /* WORKSPACE */:
                return core_1.isWorkspaceIdentifier(workspaceIdentifier) && this.workspace.id === workspaceIdentifier.id;
        }
        return false;
    }
    extAcceptChangedWorkbenchState(workspace, event) {
        this.workspace = this.reviveWorkspaceUris(workspace);
        this._onDidChangeWorkbenchState.fire(event);
        this.releaseUpdateWorkspaceBarrier();
    }
    extAcceptChangedWorkspaceFolders(workspace, event) {
        this.workspace = this.reviveWorkspaceUris(workspace);
        this._onDidChangeWorkspaceFolders.fire(event);
        this.releaseUpdateWorkspaceBarrier();
    }
    extAcceptChangedWorkspaceName(workspace) {
        this.workspace = this.reviveWorkspaceUris(workspace);
        this._onDidChangeWorkspaceName.fire();
        this.releaseUpdateWorkspaceBarrier();
    }
    releaseWorkspaceBarrier() {
        if (!this.completeWorkspaceBarrier.isOpen()) {
            this.completeWorkspaceBarrier.open();
        }
    }
    releaseUpdateWorkspaceBarrier() {
        if (this.updateWorkspaceBarrier && !this.updateWorkspaceBarrier.isOpen()) {
            this.updateWorkspaceBarrier.open();
        }
    }
    reviveWorkspaceUris(workspaceData) {
        let workspaceConfig, workspaceFolders;
        if (workspaceData.folders) {
            workspaceFolders = workspaceData.folders.map(({ index, uri: uriComponent, name }) => (new core_1.WorkspaceFolder({ index, uri: uri_1.URI.revive(uriComponent), name })));
        }
        if (workspaceData.configuration) {
            workspaceConfig = uri_1.URI.revive(workspaceData.configuration);
        }
        return new core_1.Workspace(workspaceData.id, workspaceFolders, workspaceConfig);
    }
};
AbstractWorkspaceContextService = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(core_2.CommonType.LOGGER)),
    __metadata("design:paramtypes", [Object])
], AbstractWorkspaceContextService);
exports.AbstractWorkspaceContextService = AbstractWorkspaceContextService;
