// @workbench-stack/platform-server/lib/services/configuration/__tests__/mock-configuration-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockConfigurationService = void 0;
const core_1 = require("@workbench-stack/core");
/**
 * TODO: Move to workspace module
 */
class MockConfigurationService {
    constructor(configuration = {}) {
        this.configuration = configuration;
    }
    inspect(key, overrides) {
        return {
            value: core_1.getConfigurationValue(this.getValue(), key),
            default: core_1.getConfigurationValue(this.getValue(), key),
            user: core_1.getConfigurationValue(this.getValue(), key),
            workspaceFolder: void 0, folder: void 0,
        };
    }
    keys() { return { default: [], user: [], workspace: [], workspaceFolder: [] }; }
    getValue(value) {
        console.log('MOCK CONFOIG-getVALUE');
        if (!value) {
            return this.configuration;
        }
        const valuePath = value.split('.');
        let object = this.configuration;
        while (valuePath.length && object) {
            object = object[valuePath.shift()];
        }
        return object;
    }
    updateValue() { return null; }
    getConfigurationData() { return null; }
    onDidChangeConfiguration() { return { dispose() { } }; }
    reloadConfiguration() { return null; }
    dispose() { }
    ;
}
exports.MockConfigurationService = MockConfigurationService;
