// @workbench-stack/core/lib/core/actions/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuId = exports.isISubmenuItem = exports.isMenutItem = void 0;
function isMenutItem(item) {
    return item.command !== undefined;
}
exports.isMenutItem = isMenutItem;
function isISubmenuItem(item) {
    return item.submenu !== undefined;
}
exports.isISubmenuItem = isISubmenuItem;
var MenuId;
(function (MenuId) {
    MenuId[MenuId["CommandPalette"] = 0] = "CommandPalette";
    MenuId[MenuId["DebugBreakpointsContext"] = 1] = "DebugBreakpointsContext";
    MenuId[MenuId["DebugCallStackContext"] = 2] = "DebugCallStackContext";
    MenuId[MenuId["DebugConsoleContext"] = 3] = "DebugConsoleContext";
    MenuId[MenuId["DebugVariablesContext"] = 4] = "DebugVariablesContext";
    MenuId[MenuId["DebugWatchContext"] = 5] = "DebugWatchContext";
    MenuId[MenuId["DebugTooBar"] = 6] = "DebugTooBar";
    MenuId[MenuId["EditorContext"] = 7] = "EditorContext";
    MenuId[MenuId["EditorTitle"] = 8] = "EditorTitle";
    MenuId[MenuId["EditorTitleContext"] = 9] = "EditorTitleContext";
    MenuId[MenuId["EmptyEditorGroupContext"] = 10] = "EmptyEditorGroupContext";
    MenuId[MenuId["ExplorerContext"] = 11] = "ExplorerContext";
    MenuId[MenuId["MenubarAppearanceMenu"] = 12] = "MenubarAppearanceMenu";
    MenuId[MenuId["MenubarDebugMenu"] = 13] = "MenubarDebugMenu";
    MenuId[MenuId["MenubarEditMenu"] = 14] = "MenubarEditMenu";
    MenuId[MenuId["MenubarFileMenu"] = 15] = "MenubarFileMenu";
    MenuId[MenuId["MenubarGoMenu"] = 16] = "MenubarGoMenu";
    MenuId[MenuId["MenubarNewBreakpointMenu"] = 17] = "MenubarNewBreakpointMenu";
    MenuId[MenuId["MenubarPreferencesMenu"] = 18] = "MenubarPreferencesMenu";
    MenuId[MenuId["MenubarRecentMenu"] = 19] = "MenubarRecentMenu";
    MenuId[MenuId["MenubarSelectionMenu"] = 20] = "MenubarSelectionMenu";
    MenuId[MenuId["MenubarSwitchEditorMenu"] = 21] = "MenubarSwitchEditorMenu";
    MenuId[MenuId["MenubarSwitchGroupMenu"] = 22] = "MenubarSwitchGroupMenu";
    MenuId[MenuId["MenubarTerminalMenu"] = 23] = "MenubarTerminalMenu";
    MenuId[MenuId["MenubarViewMenu"] = 24] = "MenubarViewMenu";
    MenuId[MenuId["OpenEditorContext"] = 25] = "OpenEditorContext";
    MenuId[MenuId["ProblemsPanelContext"] = 26] = "ProblemsPanelContext";
    MenuId[MenuId["SCMChangeContext"] = 27] = "SCMChangeContext";
    MenuId[MenuId["SCMResourceContext"] = 28] = "SCMResourceContext";
    MenuId[MenuId["SCMResourceGroupContext"] = 29] = "SCMResourceGroupContext";
    MenuId[MenuId["SCMResourceControl"] = 30] = "SCMResourceControl";
    MenuId[MenuId["SCMTitle"] = 31] = "SCMTitle";
    MenuId[MenuId["SearchContext"] = 32] = "SearchContext";
    MenuId[MenuId["StatusBarWindowIndicatorMenu"] = 33] = "StatusBarWindowIndicatorMenu";
    MenuId[MenuId["TouchBarContext"] = 34] = "TouchBarContext";
    MenuId[MenuId["ViewItemContext"] = 35] = "ViewItemContext";
    MenuId[MenuId["ViewTitle"] = 36] = "ViewTitle";
    MenuId[MenuId["CommentThreadTitle"] = 37] = "CommentThreadTitle";
    MenuId[MenuId["CommentThreadActions"] = 38] = "CommentThreadActions";
    MenuId[MenuId["CommentTitle"] = 39] = "CommentTitle";
    MenuId[MenuId["CommentActions"] = 40] = "CommentActions";
})(MenuId = exports.MenuId || (exports.MenuId = {}));
