// @workbench-stack/core/lib/interfaces/extension/extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionState = exports.ExtensionType = exports.EXTENSION_IDENTIFIER_REGEX = exports.EXTENSION_IDENTIFIER_PATTERN = exports.ExtensionHostLogFileName = exports.BUILTIN_MANIFEST_CACHE_FILE = exports.USER_MANIFEST_CACHE_FILE = exports.MANIFEST_CACHE_FOLDER = void 0;
exports.MANIFEST_CACHE_FOLDER = 'CachedExtensions';
exports.USER_MANIFEST_CACHE_FILE = 'user';
exports.BUILTIN_MANIFEST_CACHE_FILE = 'builtin';
exports.ExtensionHostLogFileName = 'exthost';
// from vs/platform/extensionManagement/common/extensionManagement
exports.EXTENSION_IDENTIFIER_PATTERN = '^([a-z0-9A-Z][a-z0-9\-A-Z]*)\\.([a-z0-9A-Z][a-z0-9\-A-Z]*)$';
exports.EXTENSION_IDENTIFIER_REGEX = new RegExp(exports.EXTENSION_IDENTIFIER_PATTERN);
var ExtensionState;
(function (ExtensionState) {
    ExtensionState[ExtensionState["Installing"] = 0] = "Installing";
    ExtensionState[ExtensionState["Installed"] = 1] = "Installed";
    ExtensionState[ExtensionState["Uninstalling"] = 2] = "Uninstalling";
    ExtensionState[ExtensionState["Uninstalled"] = 3] = "Uninstalled";
})(ExtensionState = exports.ExtensionState || (exports.ExtensionState = {}));
