// @files-stack/core/lib/interfaces/files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystemProviderErrorCode = exports.FileSystemProviderCapabilities = exports.HotExitConfiguration = exports.FileChangeType = exports.FileKind = exports.FileType = void 0;
var FileType;
(function (FileType) {
    FileType[FileType["Unknown"] = 0] = "Unknown";
    FileType[FileType["File"] = 1] = "File";
    FileType[FileType["Directory"] = 2] = "Directory";
    FileType[FileType["SymbolicLink"] = 64] = "SymbolicLink";
})(FileType = exports.FileType || (exports.FileType = {}));
var FileKind;
(function (FileKind) {
    FileKind[FileKind["FILE"] = 0] = "FILE";
    FileKind[FileKind["FOLDER"] = 1] = "FOLDER";
    FileKind[FileKind["ROOT_FOLDER"] = 2] = "ROOT_FOLDER";
})(FileKind = exports.FileKind || (exports.FileKind = {}));
/**
 * Possible changes that can occur to a file.
 */
var FileChangeType;
(function (FileChangeType) {
    FileChangeType["UPDATED"] = "UPDATED";
    FileChangeType["ADDED"] = "ADDED";
    FileChangeType["DELETED"] = "DELETED";
})(FileChangeType = exports.FileChangeType || (exports.FileChangeType = {}));
// @autogenerated
// export enum FileOperationResult
// // @sri modified
// export enum AutoSaveConfigurationType {
//   OFF = 'off',
//   AFTER_DELAY = 'afterDelay',
//   ON_FOCUS_CHANGE = 'onFocusChange',
//   ON_WINDOW_CHANGE = 'onWindowChange'
// };
// @sri modified
var HotExitConfiguration;
(function (HotExitConfiguration) {
    HotExitConfiguration["OFF"] = "off";
    HotExitConfiguration["ON_EXIT"] = "onExit";
    HotExitConfiguration["ON_EXIT_AND_WINDOW_CLOSE"] = "onExitAndWindowClose";
})(HotExitConfiguration = exports.HotExitConfiguration || (exports.HotExitConfiguration = {}));
;
var FileSystemProviderCapabilities;
(function (FileSystemProviderCapabilities) {
    FileSystemProviderCapabilities[FileSystemProviderCapabilities["FileReadWrite"] = 2] = "FileReadWrite";
    FileSystemProviderCapabilities[FileSystemProviderCapabilities["FileOpenReadWriteClose"] = 4] = "FileOpenReadWriteClose";
    FileSystemProviderCapabilities[FileSystemProviderCapabilities["FileFolderCopy"] = 8] = "FileFolderCopy";
    FileSystemProviderCapabilities[FileSystemProviderCapabilities["PathCaseSensitive"] = 1024] = "PathCaseSensitive";
    FileSystemProviderCapabilities[FileSystemProviderCapabilities["Readonly"] = 2048] = "Readonly";
    FileSystemProviderCapabilities[FileSystemProviderCapabilities["Trash"] = 4096] = "Trash";
})(FileSystemProviderCapabilities = exports.FileSystemProviderCapabilities || (exports.FileSystemProviderCapabilities = {}));
var FileSystemProviderErrorCode;
(function (FileSystemProviderErrorCode) {
    FileSystemProviderErrorCode["FileExists"] = "EntryExists";
    FileSystemProviderErrorCode["FileNotFound"] = "EntryNotFound";
    FileSystemProviderErrorCode["FileNotADirectory"] = "EntryNotADirectory";
    FileSystemProviderErrorCode["FileIsADirectory"] = "EntryIsADirectory";
    FileSystemProviderErrorCode["NoPermissions"] = "NoPermissions";
    FileSystemProviderErrorCode["Unavailable"] = "Unavailable";
    FileSystemProviderErrorCode["Unknown"] = "Unknown";
})(FileSystemProviderErrorCode = exports.FileSystemProviderErrorCode || (exports.FileSystemProviderErrorCode = {}));
