// @files-stack/core/lib/interfaces/textfiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileContentStatus = exports.LoadReason = exports.SaveReason = exports.AutoSaveMode = exports.StateChange = exports.ModelState = void 0;
/**
 * States the text file editor model can be in.
 */
var ModelState;
(function (ModelState) {
    ModelState[ModelState["SAVED"] = 0] = "SAVED";
    ModelState[ModelState["DIRTY"] = 1] = "DIRTY";
    ModelState[ModelState["PENDING_SAVE"] = 2] = "PENDING_SAVE";
    /**
     * A model is in conflict mode when changes cannot be saved because the
     * underlying file has changed. Models in conflict mode are always dirty.
     */
    ModelState[ModelState["CONFLICT"] = 3] = "CONFLICT";
    /**
     * A model is in orphan state when the underlying file has been deleted.
     */
    ModelState[ModelState["ORPHAN"] = 4] = "ORPHAN";
    /**
     * Any error that happens during a save that is not causing the CONFLICT state.
     * Models in error mode are always diry.
     */
    ModelState[ModelState["ERROR"] = 5] = "ERROR";
})(ModelState = exports.ModelState || (exports.ModelState = {}));
var StateChange;
(function (StateChange) {
    StateChange[StateChange["DIRTY"] = 0] = "DIRTY";
    StateChange[StateChange["SAVING"] = 1] = "SAVING";
    StateChange[StateChange["SAVE_ERROR"] = 2] = "SAVE_ERROR";
    StateChange[StateChange["SAVED"] = 3] = "SAVED";
    StateChange[StateChange["REVERTED"] = 4] = "REVERTED";
    StateChange[StateChange["ENCODING"] = 5] = "ENCODING";
    StateChange[StateChange["CONTENT_CHANGE"] = 6] = "CONTENT_CHANGE";
    StateChange[StateChange["ORPHANED_CHANGE"] = 7] = "ORPHANED_CHANGE";
})(StateChange = exports.StateChange || (exports.StateChange = {}));
var AutoSaveMode;
(function (AutoSaveMode) {
    AutoSaveMode[AutoSaveMode["OFF"] = 0] = "OFF";
    AutoSaveMode[AutoSaveMode["AFTER_SHORT_DELAY"] = 1] = "AFTER_SHORT_DELAY";
    AutoSaveMode[AutoSaveMode["AFTER_LONG_DELAY"] = 2] = "AFTER_LONG_DELAY";
    AutoSaveMode[AutoSaveMode["ON_FOCUS_CHANGE"] = 3] = "ON_FOCUS_CHANGE";
    AutoSaveMode[AutoSaveMode["ON_WINDOW_CHANGE"] = 4] = "ON_WINDOW_CHANGE";
})(AutoSaveMode = exports.AutoSaveMode || (exports.AutoSaveMode = {}));
var SaveReason;
(function (SaveReason) {
    SaveReason[SaveReason["EXPLICIT"] = 1] = "EXPLICIT";
    SaveReason[SaveReason["AUTO"] = 2] = "AUTO";
    SaveReason[SaveReason["FOCUS_CHANGE"] = 3] = "FOCUS_CHANGE";
    SaveReason[SaveReason["WINDOW_CHANGE"] = 4] = "WINDOW_CHANGE";
})(SaveReason = exports.SaveReason || (exports.SaveReason = {}));
var LoadReason;
(function (LoadReason) {
    LoadReason[LoadReason["EDITOR"] = 1] = "EDITOR";
    LoadReason[LoadReason["REFERENCE"] = 2] = "REFERENCE";
    LoadReason[LoadReason["OTHER"] = 3] = "OTHER";
})(LoadReason = exports.LoadReason || (exports.LoadReason = {}));
// @sri added to update the status of the progressed content
var FileContentStatus;
(function (FileContentStatus) {
    FileContentStatus["PENDING"] = "PENDING";
    FileContentStatus["START"] = "START";
    FileContentStatus["IN_PROGRESS"] = "IN-PROGRESS";
    FileContentStatus["END"] = "END";
    FileContentStatus["ERROR"] = "ERROR";
    FileContentStatus["RESTORE"] = "RESTORE";
})(FileContentStatus = exports.FileContentStatus || (exports.FileContentStatus = {}));
