// @workbench-stack/platform-server/lib/core/log/delegated-log-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegatedLogService = void 0;
const lifecycle_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/lifecycle");
class DelegatedLogService extends lifecycle_1.Disposable {
    constructor(logService) {
        super();
        this.logService = logService;
        this._register(logService);
    }
    get onDidChangeLogLevel() {
        return this.logService.onDidChangeLogLevel;
    }
    setLevel(level) {
        this.logService.setLevel(level);
    }
    getLevel() {
        return this.logService.getLevel();
    }
    trace(message, ...args) {
        this.logService.trace(message, ...args);
    }
    debug(message, ...args) {
        this.logService.debug(message, ...args);
    }
    info(message, ...args) {
        this.logService.info(message, ...args);
    }
    warn(message, ...args) {
        this.logService.warn(message, ...args);
    }
    error(message, ...args) {
        this.logService.error(message, ...args);
    }
    critical(message, ...args) {
        this.logService.critical(message, ...args);
    }
}
exports.DelegatedLogService = DelegatedLogService;
