// @workbench-stack/platform-server/lib/services/lifecycle/__tests__/lifecycle-service-sample.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestLifecycleService = void 0;
/* tslint:disable */
const event_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/event");
const inversify_1 = require("inversify");
let TestLifecycleService = class TestLifecycleService {
    constructor() {
        this._onWillShutdown = new event_1.Emitter();
        this._onShutdown = new event_1.Emitter();
    }
    when() {
        return Promise.resolve();
        ;
    }
    fireShutdown(reason = 2 /* QUIT */) {
        this._onShutdown.fire();
    }
    fireWillShutdown(event) {
        this._onWillShutdown.fire(event);
    }
    get onWillShutdown() {
        return this._onWillShutdown.event;
    }
    get onShutdown() {
        return this._onShutdown.event;
    }
};
TestLifecycleService = __decorate([
    inversify_1.injectable()
], TestLifecycleService);
exports.TestLifecycleService = TestLifecycleService;
