// @workbench-stack/platform-server/lib/services/configuration/configuration-client-moleculerservice.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationClientMoleculerService = void 0;
const moleculer_1 = require("moleculer");
const core_1 = require("@workbench-stack/core");
const async_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/async");
class ConfigurationClientMoleculerService extends moleculer_1.Service {
    constructor(broker, { container, settings }) {
        super(broker);
        this.container = container;
        this.startupBarrier = new async_1.Barrier();
        this.parseServiceSchema({
            name: "ConfigurationClientService" /* ConfigurationClientService */,
            dependencies: [
                { name: "ConfigurationService" /* ConfigurationService */ },
            ],
            started: () => __awaiter(this, void 0, void 0, function* () {
                const lifecycle = this.container.get(core_1.TYPES.ILifecycleService);
                lifecycle.when(2 /* Ready */).then(() => {
                    this.startupBarrier.open();
                });
            }),
            events: {
                ["onDidChangeConfiguration" /* onDidChangeConfiguration */]: {
                    handler: (ctx) => __awaiter(this, void 0, void 0, function* () {
                        (yield this.configurationService).extAcceptConfigurationChanged(ctx.params.data, ctx.params.event);
                    }),
                },
            },
            stopped: () => __awaiter(this, void 0, void 0, function* () {
                if (!this.startupBarrier.isOpen()) {
                    this.startupBarrier.open();
                }
                (yield this.configurationService).dispose();
            }),
        });
    }
    get configurationService() {
        if (this._configurationService) {
            return this._configurationService;
        }
        return this.startupBarrier.wait().then(() => {
            return this._configurationService = this.container.get(core_1.TYPES.IConfigurationService);
        });
    }
}
exports.ConfigurationClientMoleculerService = ConfigurationClientMoleculerService;
