// @workbench-stack/core/lib/core/preferences/preferences.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreDefinedSettingsURI = exports.getSettingsTargetName = void 0;
const nls_1 = require("@vscode-alt/monaco-editor/esm/vs/nls");
const uri_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/uri");
const network_1 = require("../network");
function getSettingsTargetName(target, resource, workspaceContextService) {
    switch (target) {
        case 1 /* USER */:
        case 2 /* USER_LOCAL */:
            return nls_1.localize('userSettingsTarget', "User Settings");
        case 4 /* WORKSPACE */:
            return nls_1.localize('workspaceSettingsTarget', "Workspace Settings");
        case 5 /* WORKSPACE_FOLDER */:
            const folder = workspaceContextService.getWorkspaceFolder(resource);
            return folder ? folder.name : '';
    }
    return '';
}
exports.getSettingsTargetName = getSettingsTargetName;
const DefaultURI = uri_1.URI.from({ scheme: network_1.Schemas.cdecode, authority: 'defaultsettings', path: '/settings.json' });
const UserURI = uri_1.URI.from({ scheme: "mongo", authority: 'defaultsettings', path: '/settings.json' });
const WorkspaceURI = uri_1.URI.from({ scheme: network_1.Schemas.cdecode, authority: 'defaultsettings', path: '/workspaceSettings.json' });
const WorkspaceFolderURI = uri_1.URI.from({ scheme: network_1.Schemas.cdecode, authority: 'defaultsettings', path: '/resourceSettings.json' });
exports.PreDefinedSettingsURI = {
    Default: DefaultURI,
    User: UserURI,
    Workspace: WorkspaceURI,
    WorkspaceFolder: WorkspaceFolderURI,
};
