// @workbench-stack/core/lib/core/editor/editor-input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateResourceEditorInput = exports.generateDiffEditorInput = exports.generateSideBySideEditorInput = void 0;
const interfaces_1 = require("../../interfaces");
function generateSideBySideEditorInput(name, description, _details, _master) {
    return {
        label: name,
        master: _master,
        details: _details,
        typeId: "SIDE_BY_SIDE_EDITOR_INPUT" /* SIDE_BY_SIDE_EDITOR_INPUT */,
        description: description,
        __typename: "SideBySideEditorInput" /* SideBySideEditorInput */,
    };
}
exports.generateSideBySideEditorInput = generateSideBySideEditorInput;
function generateDiffEditorInput(name, description, original, modified, forceOpenAsBinary) {
    return {
        label: name,
        typeId: "DIFF_EDITOR_INPUT" /* DIFF_EDITOR_INPUT */,
        originalInput: original,
        modifiedInput: modified,
        preferredEditorId: forceOpenAsBinary ? interfaces_1.BINARY_DIFF_EDITOR_ID : interfaces_1.TEXT_DIFF_EDITOR_ID,
        __typename: "DiffEditorInput" /* DiffEditorInput */,
    };
}
exports.generateDiffEditorInput = generateDiffEditorInput;
function generateResourceEditorInput(name, description, resource, preferredMode) {
    return {
        name,
        description,
        resource,
        __typename: "ResourceEditorInput" /* ResourceEditorInput */,
    };
}
exports.generateResourceEditorInput = generateResourceEditorInput;
