// @workbench-stack/core/lib/core/severity/severity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Severity = void 0;
const strings = require("@vscode-alt/monaco-editor/esm/vs/base/common/strings");
var Severity;
(function (Severity) {
    Severity["Ignore"] = "Ignore";
    Severity["Info"] = "Info";
    Severity["Warning"] = "Warning";
    Severity["Error"] = "Error";
})(Severity || (Severity = {}));
exports.Severity = Severity;
(function (Severity) {
    const _error = 'error';
    const _warning = 'warning';
    const _warn = 'warn';
    const _info = 'info';
    /**
     * Parses 'error', 'warning', 'warn', 'info' in call casings
     * and falls back to ignore.
     */
    function fromValue(value) {
        if (!value) {
            return Severity.Ignore;
        }
        if (strings.equalsIgnoreCase(_error, value)) {
            return Severity.Error;
        }
        if (strings.equalsIgnoreCase(_warning, value) || strings.equalsIgnoreCase(_warn, value)) {
            return Severity.Warning;
        }
        if (strings.equalsIgnoreCase(_info, value)) {
            return Severity.Info;
        }
        return Severity.Ignore;
    }
    Severity.fromValue = fromValue;
})(Severity || (Severity = {}));
exports.Severity = Severity;
