// @workbench-stack/platform-server/lib/services/configuration/configuration-moleculerservice.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationMoleculerService = void 0;
const moleculer_1 = require("moleculer");
const core_1 = require("@workbench-stack/core");
class ConfigurationMoleculerService extends moleculer_1.Service {
    constructor(broker, { container, settings }) {
        super(broker);
        const topic = "ConfigurationService" /* ConfigurationService */;
        this.parseServiceSchema({
            name: topic,
            started: () => __awaiter(this, void 0, void 0, function* () {
                this._configuration = container.get(core_1.TYPES.IConfigurationService);
            }),
            actions: {
                ["getConfigurationData" /* getConfigurationData */]: {
                    handler: (ctx) => __awaiter(this, void 0, void 0, function* () {
                        return yield this._configuration.getConfigurationData();
                    }),
                    cache: true,
                },
                ["updateValue" /* updateValue */]: {
                    handler: (ctx) => __awaiter(this, void 0, void 0, function* () {
                        return yield this._configuration.updateValue(ctx.params.key, ctx.params.value, ctx.params.overrides, ctx.params.target, ctx.params.donotNotifyError);
                    }),
                },
                ["reloadConfiguration" /* reloadConfiguration */]: {
                    handler: (ctx) => __awaiter(this, void 0, void 0, function* () {
                        return yield this._configuration.reloadConfiguration();
                    }),
                },
            },
            stopped: () => __awaiter(this, void 0, void 0, function* () {
                return this._configuration.dispose();
            }),
        });
    }
}
exports.ConfigurationMoleculerService = ConfigurationMoleculerService;
