// @workbench-stack/platform-server/lib/core/workspace/configurations/all-keys-configuration-change-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllKeysConfigurationChangeEvent = void 0;
/* tslint:disable */
const core_1 = require("@workbench-stack/core");
const map_1 = require("@vscode-alt/monaco-editor/esm/vs/base/common/map");
class AllKeysConfigurationChangeEvent extends core_1.AbstractConfigurationChangeEvent {
    constructor(_configuration, source, sourceConfig) {
        super();
        this._configuration = _configuration;
        this.source = source;
        this.sourceConfig = sourceConfig;
        this._changedConfiguration = null;
    }
    get changedConfiguration() {
        if (!this._changedConfiguration) {
            this._changedConfiguration = new core_1.ConfigurationModel();
            this.updateKeys(this._changedConfiguration, this.affectedKeys);
        }
        return this._changedConfiguration;
    }
    get changedConfigurationByResource() {
        return new map_1.ResourceMap();
    }
    get affectedKeys() {
        return this._configuration.allKeys();
    }
    affectsConfiguration(config, resource) {
        return this.doesConfigurationContains(this.changedConfiguration, config);
    }
}
exports.AllKeysConfigurationChangeEvent = AllKeysConfigurationChangeEvent;
