// @cdm-logger/core/lib/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLoggerOptions = exports.makeLogger = exports.getSettingsLevel = void 0;
function getSettingsLevel(settings) {
    return settings.level || "info";
}
exports.getSettingsLevel = getSettingsLevel;
function makeLogger(logger, name, ...streams) {
    const logName = typeof name === "object" ? name.constructor.toString().match(/class ([\w|_]+)/)[1] : name;
    return logger.createLogger(getLoggerOptions(logName, ...streams));
}
exports.makeLogger = makeLogger;
function getLoggerOptions(name, ...streams) {
    if (!name) {
        throw Error("Cannot create LoggerOptions without a log name");
    }
    const options = {
        name: name,
        src: true,
        serializers: {
            err: err => {
                return {
                    message: err.message,
                    name: err.name,
                    stack: err.stack,
                    code: err.code,
                    signal: err.signal,
                    internalMessage: err.internalMessage,
                    errors: err.errors
                };
            }
        }
    };
    if (streams && streams.length) {
        options.streams = streams;
    }
    return options;
    ;
}
exports.getLoggerOptions = getLoggerOptions;
